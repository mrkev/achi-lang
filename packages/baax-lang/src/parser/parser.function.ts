import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { sublang } from "./sublang";
import { Node, withAt } from "./Node";

export type LangType_Function = {
  // printPoint(x: 3, y: 4)
  FunctionCall: Node<{
    kind: "FunctionCall";
    identifier: LangType["ValueIdentifier"];
    argument: LangType["RecordLiteral"];
  }>;

  // FunctionDefinition: {
  //   kind: "FunctionDefinition";
  //   argument: LangType["RecordDefinition"];
  //   body: LangType["Block"];
  // };

  // (x: number) => { ... }
  AnonymousFunctionLiteral: Node<{
    kind: "AnonymousFunctionLiteral";
    argument: LangType["RecordDefinition"];
    block: LangType["Block"];
  }>;
};

/**
 * Defines binary operations
 */
export const LangDef_Function = sublang<LangType, LangType_Function>({
  // log(msg: "hello")
  FunctionCall: (r) => {
    return withAt(
      Parsimmon.seqMap(
        r.ValueIdentifier,
        r.RecordLiteral,
        function (identifier, argument) {
          return {
            kind: "FunctionCall",
            identifier,
            argument,
          };
        }
      )
    );
  },

  // (foo: string) => {}
  AnonymousFunctionLiteral: (r) => {
    return withAt(
      Parsimmon.seqMap(
        r.RecordDefinition,
        r.__,
        Parsimmon.string("=>"),
        r.__,
        r.Block,
        (argument, _1, _2, _3, block) => {
          return {
            kind: "AnonymousFunctionLiteral",
            argument,
            block,
          };
        }
      )
    );
  },

  /* 
  TODO;
  const add = (x: number, y: number) => number {
  }

  function foo(x: number): number {
    return x + 3;
  }
   */
  // FunctionDefinition: (r) => {
  //   return Parsimmon.seqMap(
  //     r.RecordDefinition,
  //     r._,
  //     Parsimmon.string("=>"),
  //     // TODO: type tag?
  //     r._,
  //     r.Block,
  //     function (argument, _1, _2, _3, body) {
  //       return {
  //         kind: "FunctionDefinition",
  //         argument,
  //         body,
  //       };
  //     }
  //   );
  // },
});
