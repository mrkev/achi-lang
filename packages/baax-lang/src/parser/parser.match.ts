import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { sublang, SublangTypedRule } from "./sublang";
import { Node, withAt } from "./Node";

export type LangType_Match = {
  CaseEntry: Node<{
    kind: "CaseEntry";
    guard: LangType["Expression"];
    block: LangType["Block"];
  }>;

  MatchFunction: Node<{
    kind: "MatchFunction";
    identifier: LangType["ValueIdentifier"];
    matchType: LangType["TypeExpression"];
    block: LangType["BlockOfCases"];
  }>;

  MatchExpression: Node<{
    kind: "MatchExpression";
    expression: LangType["Expression"];
    block: LangType["BlockOfCases"];
  }>;

  BlockOfCases: Node<{
    kind: "BlockOfCases";
    caseEntries: Array<LangType["CaseEntry"]>;
  }>;
};

/**
 * Defines binary operations
 */
export const LangDef_Match: SublangTypedRule<LangType, LangType_Match> =
  sublang<LangType, LangType_Match>({
    // case (card: King(value)): { ... }
    CaseEntry: (r) => {
      return withAt(
        Parsimmon.seqMap(
          Parsimmon.string("case"),
          r.__,
          r.Expression,
          r._,
          Parsimmon.string(":"),
          r.__,
          r.Block,
          function (_0, _1, expression, _3, _4, _5, block) {
            return {
              kind: "CaseEntry",
              guard: expression,
              block,
            };
          }
        )
      );
    },

    // function foo matches Type {}
    MatchFunction: (r) => {
      return withAt(
        Parsimmon.seqMap(
          Parsimmon.string("function"),
          r.__,
          r.ValueIdentifier,
          r.__,
          Parsimmon.string("matches"),
          r.__,
          r.TypeExpression,
          r._,
          r.BlockOfCases,
          function (_0, _1, identifier, _3, _4, _5, type, _7, block) {
            return {
              kind: "MatchFunction",
              identifier,
              matchType: type,
              block,
            };
          }
        )
      );
    },

    // match (card) { case King(): ... }
    MatchExpression: (r) => {
      return withAt(
        Parsimmon.seqMap(
          Parsimmon.string("match"),
          r._,
          Parsimmon.string("("),
          r._,
          r.Expression,
          r._,
          Parsimmon.string(")"),
          r._,
          r.BlockOfCases,
          function (_0, _1, _2, _3, expression, _5, _6, _7, block) {
            return {
              kind: "MatchExpression",
              expression,
              block,
            };
          }
        )
      );
    },

    BlockOfCases: (r) => {
      const caseListParser = Parsimmon.sepBy(r.CaseEntry, r.__nl).trim(r._nl);
      return withAt(
        Parsimmon.string("{")
          // TBD: sepBy1 and make a different parser for Unit?
          .then(caseListParser)
          .skip(Parsimmon.string("}"))
          .map((caseEntries) => {
            return {
              kind: "BlockOfCases",
              caseEntries,
            };
          })
      );
    },
  });
