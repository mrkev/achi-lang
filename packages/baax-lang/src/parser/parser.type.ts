import * as Parsimmon from "parsimmon";
import { Node } from "./Node";
import { LangType } from "./parser";
import { sublang } from "./sublang";

type PrefixUnaryTypeOperator =
  // for numbers
  | "-"
  // to make readonly
  | "+";
export type PrefixUnaryTypeOperation = Node<{
  kind: "PrefixUnaryTypeOperation";
  operator: PrefixUnaryTypeOperator;
  value: NEXT_PARSER;
}>;

type BinaryTypeOperator = "|" | "&";
export type BinaryTypeOperation = Node<{
  kind: "BinaryTypeOperation";
  operator: BinaryTypeOperator;
  left: NEXT_PARSER;
  right: NEXT_PARSER;
}>;

export type OperatableTypeExpression =
  | LangType["NumberLiteral"]
  | LangType["TypeIdentifier"]
  | LangType["StringLiteral"]
  | LangType["BooleanLiteral"]
  | LangType["RecordDefinition"];

type NEXT_PARSER =
  | OperatableTypeExpression
  | PrefixUnaryTypeOperation
  | BinaryTypeOperation;

// Takes a parser for the prefix operator, and a parser for the base thing being
// parsed, and parses as many occurrences as possible of the prefix operator.
// Note that the parser is created using `Parsimmon.lazy` because it's recursive. It's
// valid for there to be zero occurrences of the prefix operator.
function PREFIX(
  operatorsParser: Parsimmon.Parser<OpLoc<PrefixUnaryTypeOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
): Parsimmon.Parser<NEXT_PARSER> {
  const parser = Parsimmon.lazy<NEXT_PARSER>(() => {
    const res: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.seqMap(
      Parsimmon.index,
      operatorsParser,
      parser,
      Parsimmon.index,
      (start, operator, value, end) =>
        ({
          kind: "PrefixUnaryTypeOperation",
          operator: operator.op,
          value,
          "@": { start, end },
        } as const)
    ).or(nextParser);
    return res;
  });
  return parser;
}

type OpLoc<T extends string> = {
  op: T;
  start: Parsimmon.Index;
  end: Parsimmon.Index;
};

// Takes a parser for all the operators at this precedence level, and a parser
// that parsers everything at the next precedence level, and returns a parser
// that parses as many binary operations as possible, associating them to the
// left. (e.g. 1-2-3 is (1-2)-3 not 1-(2-3))
function BINARY_LEFT(
  operatorsParser: Parsimmon.Parser<OpLoc<BinaryTypeOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
) {
  // We run into a similar problem as with the `POSTFIX` parser above where we
  // can't recurse in the direction we want, so we have to resort to parsing an
  // entire list of operator chunks and then using `.reduce` to manually nest
  // them again.
  //
  // Example:
  //
  // INPUT  :: "1+2+3"
  // PARSE  :: [1, ["+", 2], ["+", 3]]
  // REDUCE :: ["+", ["+", 1, 2], 3]
  return Parsimmon.seqMap(
    nextParser,
    Parsimmon.seq(operatorsParser, nextParser).many(),
    (first, rest) => {
      return rest.reduce((acc, ch) => {
        const [op, another] = ch;
        return {
          kind: "BinaryTypeOperation",
          operator: op.op,
          left: acc,
          right: another,
          "@": {
            start: acc["@"].start,
            end: another["@"].end,
          },
        } as const;
      }, first);
    }
  );
}

export function OperatorParser(r: Parsimmon.TypedLanguage<LangType>) {
  // Operators should allow whitespace around them, but not require it. This
  // helper combines multiple operators together with names.
  //
  // Example: operators({Add: "+", Sub: "-"})
  //
  // Gives back an operator that parses either + or - surrounded by optional
  // whitespace, and gives back the word "Add" or "Sub" instead of the character.
  function operators<Ops extends readonly string[]>(
    ops: Ops
  ): Parsimmon.Parser<OpLoc<Ops[number]>> {
    const withWhitespace = ops.map((sym) =>
      Parsimmon.seqMap(
        Parsimmon.index,
        r._,
        Parsimmon.string(sym),
        r._,
        Parsimmon.index,
        (start, _0, op, _2, end) => {
          return { op, start, end };
        }
      )
    );
    const res = Parsimmon.alt(...withWhitespace);
    return res;
  }

  // A basic value is any parenthesized expression or a number.
  // TODO: start, end don't take the parens into consideration. Is that an issue?
  const Basic: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.lazy<NEXT_PARSER>(() =>
    Parsimmon.string("(")
      .then(MyMath)
      .skip(Parsimmon.string(")"))
      .or(r.RecordDefinition)
      .or(r.NumberLiteral)
      .or(r.StringLiteral)
      .or(r.BooleanLiteral)
      .or(r.TypeIdentifier)
  );

  // Now we can describe the operators in order by precedence. You just need to
  // re-order the table.
  // const table = [
  //   { type: POSTFIX, ops: operators(["!"] as const) },
  //   { type: PREFIX, ops: operators(["-", "!"] as const) },
  //   { type: BINARY_RIGHT, ops: operators(["^"] as const) },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["*", "/"] as const),
  //   },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["+", "-"] as const),
  //   },
  // ];

  const table: Array<
    (next: Parsimmon.Parser<NEXT_PARSER>) => Parsimmon.Parser<NEXT_PARSER>
  > = [
    (next) => PREFIX(operators(["-", "+"] as const), next),
    (next) => BINARY_LEFT(operators(["&", "|"] as const), next),
  ];

  // Start off with Num as the base parser for numbers and thread that through the
  // entire table of operator parsers.
  const tableParser = table.reduce((next, level) => level(next), Basic);

  // The above is equivalent to:
  //
  // TYPE(operators({...}),
  //   TYPE(operators({...}),
  //     TYPE(operators({...})),
  //       TYPE(operators({...}),
  //         TYPE(operators({...}), ...))))
  //
  // But it's easier if to see what's going on and reorder the precedence if we
  // keep it in a table instead of nesting it all manually.

  // This is our version of a math expression.
  const MyMath = tableParser.trim(r._);
  return MyMath;
}

export type LangType_Type = {
  TypeExpression:
    | OperatableTypeExpression
    | PrefixUnaryTypeOperation
    | BinaryTypeOperation;

  // :string
  TypeTag: Node<{
    kind: "TypeTag";
    optional: boolean;
    typeExpression: LangType["TypeExpression"];
  }>;
};

/**
 * Defines type expressions and operations
 */
export const LangDef_Type = sublang<LangType, LangType_Type>({
  TypeExpression: OperatorParser,

  // :string
  TypeTag: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.index,
      Parsimmon.alt(Parsimmon.string("?:"), Parsimmon.string(":")),
      r._,
      r.TypeExpression,
      Parsimmon.index,
      (start, tagger, _3, typeExpression, end) => {
        const optional = tagger === "?:";
        return {
          kind: "TypeTag",
          optional,
          typeExpression,
          "@": { start, end },
        };
      }
    );
  },
});
