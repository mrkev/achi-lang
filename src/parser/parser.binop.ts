import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { MathParser, NEXT_PARSER } from "./parser.binex";
import { sublang } from "./sublang";

// expression =
//     term
//     | expression "+" term
//     | expression "-" term .
// term =
//     factor
//     | term "*" factor
//     | term "/" factor .
// factor =
//     number
//     | "(" expression ")" .

export type BinaryExpression = {
  kind: "BinaryExpression";
  operator: string;
  left: LangType["Expression"];
  right: LangType["Expression"];
};

export type LangType_BinOp = {
  // BinaryOperator: string;
  // BinaryOperation: {
  //   kind: "BinaryOperation";
  //   operator: string;
  //   left: LangType["Expression"];
  //   right: LangType["Expression"];
  // };

  // BinaryExpression_Expr
  // BinaryExpression: BinaryExpression | LangType["Expression"];
  // BinaryExpression_Term: BinaryExpression | LangType["Expression"];
  // BinaryExpression_Factor: BinaryExpression | LangType["Expression"];
  OperatorExpression: NEXT_PARSER;
};

/**
 * Defines binary operations
 */
export const LangDef_BinOp = sublang<LangType, LangType_BinOp>({
  OperatorExpression: MathParser,

  // BinaryExpression: (r) => {
  //   return Parsimmon.alt(
  //     r.BinaryExpression_Term,
  //     Parsimmon.seqMap(
  //       r.BinaryExpression,
  //       r._,
  //       Parsimmon.regex(/(\+|\-)/),
  //       r._,
  //       r.BinaryExpression_Term,
  //       (expr, _1, op, _3, term): LangType["BinaryExpression"] => {
  //         return {
  //           kind: "BinaryExpression",
  //           left: expr,
  //           operator: op,
  //           right: term,
  //         };
  //       }
  //     )
  //   );
  // },

  // // // term =
  // //     factor
  // //     | term "*" factor
  // //     | term "/" factor .
  // BinaryExpression_Term: (r) => {
  //   return Parsimmon.alt(
  //     r.BinaryExpression_Factor,
  //     Parsimmon.seqMap(
  //       r.BinaryExpression_Term,
  //       r._,
  //       Parsimmon.regex(/(\*|\/)/),
  //       r._,
  //       r.BinaryExpression_Factor,
  //       (term, _1, op, _3, factor): LangType["BinaryExpression_Term"] => {
  //         return {
  //           kind: "BinaryExpression",
  //           left: term,
  //           operator: op,
  //           right: factor,
  //         };
  //       }
  //     )
  //   );
  // },

  // BinaryExpression_Factor: (r) => {
  //   return Parsimmon.seqMap(
  //     Parsimmon.string("<<"),
  //     r._,
  //     r.BinaryExpression,
  //     r._,
  //     Parsimmon.string(">>"),
  //     (_0, _1, expr, _3, _4) => {
  //       return expr;
  //     }
  //   ).or(r.NumberLiteral);
  // },

  // Arithmetic: + - / * %
  // Comparison: all double << >> == !=
  // Logic: || &&
  // Todo, bitwise?
  // BinaryOperator: () => {
  //   return Parsimmon.regex(/(\+|\-|\/|\*|\%|\=\=|\!\=|\>\>|\<\<|\|\||\&\&)/);
  // },

  // 3 + 3
  // BinaryOperation: (r) => {
  //   return Parsimmon.seqMap(
  //     r.Expression,
  //     r._,
  //     r.BinaryOperator,
  //     r._,
  //     r.Expression,
  //     (left, _0, operator, _1, right) => {
  //       return {
  //         kind: "BinaryOperation",
  //         operator,
  //         left,
  //         right,
  //       };
  //     }
  //   );
  // },
});

// Factor
// = "(" _ expr:Expression _ ")" { return expr; }
// / Integer
