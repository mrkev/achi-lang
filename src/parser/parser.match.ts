import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { sublang } from "./sublang";

export type LangType_Match = {
  CaseEntry: {
    kind: "CaseEntry";
    guard: LangType["Expression"];
    block: LangType["Block"];
  };

  MatchFunction: {
    kind: "MatchFunction";
    identifier: LangType["ValueIdentifier"];
    matchType: LangType["Type"];
    block: LangType["BlockOfCases"];
  };

  BlockOfCases: {
    kind: "BlockOfCases";
    caseEntries: Array<LangType["CaseEntry"]>;
  };
};

/**
 * Defines binary operations
 */
export const LangDef_Match = sublang<LangType, LangType_Match>({
  // case (card: King(value)): { ... }
  CaseEntry: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("case"),
      r.__,
      r.Expression,
      r._,
      Parsimmon.string(":"),
      r.__,
      r.Block,
      function (_0, _1, expression, _3, _4, _5, block) {
        return {
          kind: "CaseEntry",
          guard: expression,
          block,
        };
      }
    );
  },

  // function foo matches Type {}
  MatchFunction: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("function"),
      r.__,
      r.ValueIdentifier,
      r.__,
      Parsimmon.string("matches"),
      r.__,
      r.Type,
      r.__,
      r.BlockOfCases,
      function (_0, _1, identifier, _3, _4, _5, type, _7, block) {
        return {
          kind: "MatchFunction",
          identifier,
          matchType: type,
          block: block,
        };
      }
    );
  },

  BlockOfCases: (r) => {
    const caseListParser = Parsimmon.sepBy(r.CaseEntry, r.__nl).trim(r._nl);

    return (
      Parsimmon.string("{")
        // TBD: sepBy1 and make a different parser for Unit?
        .then(caseListParser)
        .skip(Parsimmon.string("}"))
        .map((caseEntries) => {
          return {
            kind: "BlockOfCases",
            caseEntries,
          };
        })
    );
  },
});
