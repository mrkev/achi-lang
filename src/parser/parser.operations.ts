import { LangType } from "./parser";
import {
  BinaryOperation,
  OperatableExpression,
  OperatorParser,
  PrefixUnaryOperation,
  SuffixUnaryOperation,
} from "./parser.binex";
import { sublang } from "./sublang";

export type LangType_BinOp = {
  OperationExpression:
    | OperatableExpression
    | PrefixUnaryOperation
    | SuffixUnaryOperation
    | BinaryOperation;
};

/**
 * Defines unary and binary operations
 */
export const LangDef_BinOp = sublang<LangType, LangType_BinOp>({
  OperationExpression: OperatorParser,
});
