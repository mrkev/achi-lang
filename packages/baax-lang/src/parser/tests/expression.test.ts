import { Lang } from "../parser";

test("ALON", () => {
  const result = Lang.Expression.tryParse(`(
    animals: {
      l1: (id: "l1", kind: "lion"),
      l2: (id: "l2", kind: "lion"),
      t1: (id: "t1", kind: "tiger"),
      t2: (id: "t2", kind: "tiger")
    },
    environments: [
      "jungle", "desert", "savannah"
    ]
  )`);
  expect(result).toMatchSnapshot();
});
