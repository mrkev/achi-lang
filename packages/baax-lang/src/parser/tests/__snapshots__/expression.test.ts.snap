// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ALON 1`] = `
{
  "@": {
    "end": {
      "column": 4,
      "line": 11,
      "offset": 235,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "definitions": [
    {
      "@": {
        "end": {
          "column": 6,
          "line": 7,
          "offset": 167,
        },
        "start": {
          "column": 5,
          "line": 2,
          "offset": 6,
        },
      },
      "expression": {
        "@": {
          "end": {
            "column": 6,
            "line": 7,
            "offset": 167,
          },
          "start": {
            "column": 14,
            "line": 2,
            "offset": 15,
          },
        },
        "entries": [
          {
            "@": {
              "end": {
                "column": 35,
                "line": 3,
                "offset": 51,
              },
              "start": {
                "column": 7,
                "line": 3,
                "offset": 23,
              },
            },
            "expression": {
              "@": {
                "end": {
                  "column": 35,
                  "line": 3,
                  "offset": 51,
                },
                "start": {
                  "column": 11,
                  "line": 3,
                  "offset": 27,
                },
              },
              "definitions": [
                {
                  "@": {
                    "end": {
                      "column": 20,
                      "line": 3,
                      "offset": 36,
                    },
                    "start": {
                      "column": 12,
                      "line": 3,
                      "offset": 28,
                    },
                  },
                  "expression": {
                    "@": {
                      "end": {
                        "column": 20,
                        "line": 3,
                        "offset": 36,
                      },
                      "start": {
                        "column": 16,
                        "line": 3,
                        "offset": 32,
                      },
                    },
                    "kind": "StringLiteral",
                    "value": "l1",
                  },
                  "identifier": {
                    "@": {
                      "end": {
                        "column": 14,
                        "line": 3,
                        "offset": 30,
                      },
                      "start": {
                        "column": 12,
                        "line": 3,
                        "offset": 28,
                      },
                    },
                    "kind": "ValueIdentifier",
                    "value": "id",
                  },
                  "kind": "NamedLiteral",
                },
                {
                  "@": {
                    "end": {
                      "column": 34,
                      "line": 3,
                      "offset": 50,
                    },
                    "start": {
                      "column": 22,
                      "line": 3,
                      "offset": 38,
                    },
                  },
                  "expression": {
                    "@": {
                      "end": {
                        "column": 34,
                        "line": 3,
                        "offset": 50,
                      },
                      "start": {
                        "column": 28,
                        "line": 3,
                        "offset": 44,
                      },
                    },
                    "kind": "StringLiteral",
                    "value": "lion",
                  },
                  "identifier": {
                    "@": {
                      "end": {
                        "column": 26,
                        "line": 3,
                        "offset": 42,
                      },
                      "start": {
                        "column": 22,
                        "line": 3,
                        "offset": 38,
                      },
                    },
                    "kind": "ValueIdentifier",
                    "value": "kind",
                  },
                  "kind": "NamedLiteral",
                },
              ],
              "kind": "RecordLiteral",
            },
            "identifier": {
              "@": {
                "end": {
                  "column": 9,
                  "line": 3,
                  "offset": 25,
                },
                "start": {
                  "column": 7,
                  "line": 3,
                  "offset": 23,
                },
              },
              "kind": "ValueIdentifier",
              "value": "l1",
            },
            "kind": "NamedLiteral",
          },
          {
            "@": {
              "end": {
                "column": 35,
                "line": 4,
                "offset": 87,
              },
              "start": {
                "column": 7,
                "line": 4,
                "offset": 59,
              },
            },
            "expression": {
              "@": {
                "end": {
                  "column": 35,
                  "line": 4,
                  "offset": 87,
                },
                "start": {
                  "column": 11,
                  "line": 4,
                  "offset": 63,
                },
              },
              "definitions": [
                {
                  "@": {
                    "end": {
                      "column": 20,
                      "line": 4,
                      "offset": 72,
                    },
                    "start": {
                      "column": 12,
                      "line": 4,
                      "offset": 64,
                    },
                  },
                  "expression": {
                    "@": {
                      "end": {
                        "column": 20,
                        "line": 4,
                        "offset": 72,
                      },
                      "start": {
                        "column": 16,
                        "line": 4,
                        "offset": 68,
                      },
                    },
                    "kind": "StringLiteral",
                    "value": "l2",
                  },
                  "identifier": {
                    "@": {
                      "end": {
                        "column": 14,
                        "line": 4,
                        "offset": 66,
                      },
                      "start": {
                        "column": 12,
                        "line": 4,
                        "offset": 64,
                      },
                    },
                    "kind": "ValueIdentifier",
                    "value": "id",
                  },
                  "kind": "NamedLiteral",
                },
                {
                  "@": {
                    "end": {
                      "column": 34,
                      "line": 4,
                      "offset": 86,
                    },
                    "start": {
                      "column": 22,
                      "line": 4,
                      "offset": 74,
                    },
                  },
                  "expression": {
                    "@": {
                      "end": {
                        "column": 34,
                        "line": 4,
                        "offset": 86,
                      },
                      "start": {
                        "column": 28,
                        "line": 4,
                        "offset": 80,
                      },
                    },
                    "kind": "StringLiteral",
                    "value": "lion",
                  },
                  "identifier": {
                    "@": {
                      "end": {
                        "column": 26,
                        "line": 4,
                        "offset": 78,
                      },
                      "start": {
                        "column": 22,
                        "line": 4,
                        "offset": 74,
                      },
                    },
                    "kind": "ValueIdentifier",
                    "value": "kind",
                  },
                  "kind": "NamedLiteral",
                },
              ],
              "kind": "RecordLiteral",
            },
            "identifier": {
              "@": {
                "end": {
                  "column": 9,
                  "line": 4,
                  "offset": 61,
                },
                "start": {
                  "column": 7,
                  "line": 4,
                  "offset": 59,
                },
              },
              "kind": "ValueIdentifier",
              "value": "l2",
            },
            "kind": "NamedLiteral",
          },
          {
            "@": {
              "end": {
                "column": 36,
                "line": 5,
                "offset": 124,
              },
              "start": {
                "column": 7,
                "line": 5,
                "offset": 95,
              },
            },
            "expression": {
              "@": {
                "end": {
                  "column": 36,
                  "line": 5,
                  "offset": 124,
                },
                "start": {
                  "column": 11,
                  "line": 5,
                  "offset": 99,
                },
              },
              "definitions": [
                {
                  "@": {
                    "end": {
                      "column": 20,
                      "line": 5,
                      "offset": 108,
                    },
                    "start": {
                      "column": 12,
                      "line": 5,
                      "offset": 100,
                    },
                  },
                  "expression": {
                    "@": {
                      "end": {
                        "column": 20,
                        "line": 5,
                        "offset": 108,
                      },
                      "start": {
                        "column": 16,
                        "line": 5,
                        "offset": 104,
                      },
                    },
                    "kind": "StringLiteral",
                    "value": "t1",
                  },
                  "identifier": {
                    "@": {
                      "end": {
                        "column": 14,
                        "line": 5,
                        "offset": 102,
                      },
                      "start": {
                        "column": 12,
                        "line": 5,
                        "offset": 100,
                      },
                    },
                    "kind": "ValueIdentifier",
                    "value": "id",
                  },
                  "kind": "NamedLiteral",
                },
                {
                  "@": {
                    "end": {
                      "column": 35,
                      "line": 5,
                      "offset": 123,
                    },
                    "start": {
                      "column": 22,
                      "line": 5,
                      "offset": 110,
                    },
                  },
                  "expression": {
                    "@": {
                      "end": {
                        "column": 35,
                        "line": 5,
                        "offset": 123,
                      },
                      "start": {
                        "column": 28,
                        "line": 5,
                        "offset": 116,
                      },
                    },
                    "kind": "StringLiteral",
                    "value": "tiger",
                  },
                  "identifier": {
                    "@": {
                      "end": {
                        "column": 26,
                        "line": 5,
                        "offset": 114,
                      },
                      "start": {
                        "column": 22,
                        "line": 5,
                        "offset": 110,
                      },
                    },
                    "kind": "ValueIdentifier",
                    "value": "kind",
                  },
                  "kind": "NamedLiteral",
                },
              ],
              "kind": "RecordLiteral",
            },
            "identifier": {
              "@": {
                "end": {
                  "column": 9,
                  "line": 5,
                  "offset": 97,
                },
                "start": {
                  "column": 7,
                  "line": 5,
                  "offset": 95,
                },
              },
              "kind": "ValueIdentifier",
              "value": "t1",
            },
            "kind": "NamedLiteral",
          },
          {
            "@": {
              "end": {
                "column": 0,
                "line": 7,
                "offset": 161,
              },
              "start": {
                "column": 7,
                "line": 6,
                "offset": 132,
              },
            },
            "expression": {
              "@": {
                "end": {
                  "column": 0,
                  "line": 7,
                  "offset": 161,
                },
                "start": {
                  "column": 11,
                  "line": 6,
                  "offset": 136,
                },
              },
              "definitions": [
                {
                  "@": {
                    "end": {
                      "column": 20,
                      "line": 6,
                      "offset": 145,
                    },
                    "start": {
                      "column": 12,
                      "line": 6,
                      "offset": 137,
                    },
                  },
                  "expression": {
                    "@": {
                      "end": {
                        "column": 20,
                        "line": 6,
                        "offset": 145,
                      },
                      "start": {
                        "column": 16,
                        "line": 6,
                        "offset": 141,
                      },
                    },
                    "kind": "StringLiteral",
                    "value": "t2",
                  },
                  "identifier": {
                    "@": {
                      "end": {
                        "column": 14,
                        "line": 6,
                        "offset": 139,
                      },
                      "start": {
                        "column": 12,
                        "line": 6,
                        "offset": 137,
                      },
                    },
                    "kind": "ValueIdentifier",
                    "value": "id",
                  },
                  "kind": "NamedLiteral",
                },
                {
                  "@": {
                    "end": {
                      "column": 35,
                      "line": 6,
                      "offset": 160,
                    },
                    "start": {
                      "column": 22,
                      "line": 6,
                      "offset": 147,
                    },
                  },
                  "expression": {
                    "@": {
                      "end": {
                        "column": 35,
                        "line": 6,
                        "offset": 160,
                      },
                      "start": {
                        "column": 28,
                        "line": 6,
                        "offset": 153,
                      },
                    },
                    "kind": "StringLiteral",
                    "value": "tiger",
                  },
                  "identifier": {
                    "@": {
                      "end": {
                        "column": 26,
                        "line": 6,
                        "offset": 151,
                      },
                      "start": {
                        "column": 22,
                        "line": 6,
                        "offset": 147,
                      },
                    },
                    "kind": "ValueIdentifier",
                    "value": "kind",
                  },
                  "kind": "NamedLiteral",
                },
              ],
              "kind": "RecordLiteral",
            },
            "identifier": {
              "@": {
                "end": {
                  "column": 9,
                  "line": 6,
                  "offset": 134,
                },
                "start": {
                  "column": 7,
                  "line": 6,
                  "offset": 132,
                },
              },
              "kind": "ValueIdentifier",
              "value": "t2",
            },
            "kind": "NamedLiteral",
          },
        ],
        "kind": "MapLiteral",
      },
      "identifier": {
        "@": {
          "end": {
            "column": 12,
            "line": 2,
            "offset": 13,
          },
          "start": {
            "column": 5,
            "line": 2,
            "offset": 6,
          },
        },
        "kind": "ValueIdentifier",
        "value": "animals",
      },
      "kind": "NamedLiteral",
    },
    {
      "@": {
        "end": {
          "column": 3,
          "line": 11,
          "offset": 234,
        },
        "start": {
          "column": 5,
          "line": 8,
          "offset": 173,
        },
      },
      "expression": {
        "@": {
          "end": {
            "column": 3,
            "line": 11,
            "offset": 234,
          },
          "start": {
            "column": 19,
            "line": 8,
            "offset": 187,
          },
        },
        "expressions": [
          {
            "@": {
              "end": {
                "column": 15,
                "line": 9,
                "offset": 203,
              },
              "start": {
                "column": 7,
                "line": 9,
                "offset": 195,
              },
            },
            "kind": "StringLiteral",
            "value": "jungle",
          },
          {
            "@": {
              "end": {
                "column": 25,
                "line": 9,
                "offset": 213,
              },
              "start": {
                "column": 17,
                "line": 9,
                "offset": 205,
              },
            },
            "kind": "StringLiteral",
            "value": "desert",
          },
          {
            "@": {
              "end": {
                "column": 5,
                "line": 10,
                "offset": 230,
              },
              "start": {
                "column": 27,
                "line": 9,
                "offset": 215,
              },
            },
            "kind": "StringLiteral",
            "value": "savannah",
          },
        ],
        "kind": "ListLiteral",
      },
      "identifier": {
        "@": {
          "end": {
            "column": 17,
            "line": 8,
            "offset": 185,
          },
          "start": {
            "column": 5,
            "line": 8,
            "offset": 173,
          },
        },
        "kind": "ValueIdentifier",
        "value": "environments",
      },
      "kind": "NamedLiteral",
    },
  ],
  "kind": "RecordLiteral",
}
`;
